pkg/pubsubsource/pubsub_test.go

```diff
@@ -154,68 +156,83 @@ func TestMain(m *testing.M) {
 		}
 		log.Fatalf("could not connect to gcloud pubsub %s", err)
 	}
+	defer pubsubClient.Close()
+	defer cancel()
 	code := m.Run()
 	if resource != nil {
 		if err := pool.Purge(resource); err != nil {
 			log.Fatalf("Couln't purge resource %s", err)
 		}
 	}
-	defer pubsubClient.Close()
 	os.Exit(code)
 }
 
 func TestPubSubSource_Read(t *testing.T) {
-
-	err := ensureTopicAndSubscription(context.TODO(), pubsubClient, TopicID, subscriptionID)
+	setupCtx, cancelSetup := context.WithCancel(context.Background())
+	err := ensureTopicAndSubscription(setupCtx, pubsubClient, TopicID, subscriptionID)
 	assert.Nil(t, err)
+	cancelSetup()
+
 	messageCh := make(chan sourcer.Message, 20)
-	//doneCh := make(chan struct{})
-	sub := pubsubClient.Subscription(subscriptionID)
-	pubsubsource := NewPubSubSource(pubsubClient, sub)
-	ctx := context.Background()
+
+	firstReadCtx, cancelFirstRead := context.WithCancel(context.Background())
+
+	subscription := pubsubClient.Subscription(subscriptionID)
+	pubsubSource := NewPubSubSource(pubsubClient, subscription)
+
+	firstSendDoneCh := make(chan struct{})
 	go func() {
-		<-time.After(5 * time.Second)
-		sendMessages(context.Background(), pubsubClient, TopicID, 5)
+		defer close(firstSendDoneCh) // Ensure to close channel after sending
+		time.Sleep(5 * time.Second)
+		sendMessages(firstReadCtx, pubsubClient, TopicID, 5)
 	}()
```

Comment:
Do we need to send messages within a goroutine? Can we make it a blocking call? It simplifies the unit tests without compromising testing the full functionalities.

Reply:
Yes because  a subscriber will only receive the message **if it is active at the time of sending messages** .If I send messages before the Read Operation then subscriber will never receive them and If I send them after ,then `context.Timeout()` happens and the Read Operation exits ,so  I have made this approach such that receive can happen parallaly (probably ) @KeranYang

Reply:
Please document the rationale. This seems related to that Pub/Sub doesn't support Pending. Please run the test multiple times locally(5+) to ensure it's stable. Thanks!

Reply:
No ,it will not miss any messages .The 6 messages sent at t2 are still in the queue ,they will be delivered to the numaflow as long as the message have not been acknowledged by another subscriber ,the messages have not yet expired based on message retention duration or the acknowledgement deadline has not passed. So in normal scenarios it wont miss the 6 messages while it was processing the previous batch